ceph/volume.go

```diff
@@ -0,0 +1,190 @@
+	defer v.m.Unlock()
+	// TODO: Might want to map with --options rw/ro here, but then we need to sneak in the RW flag somehow
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+
+	cmd := exec.Command("rbd", "create", v.Name, "--size", fmt.Sprintf("%v", sizeMB))
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err == nil {
+		log.Infof("Created Ceph volume '%s'", v.Name)
+	} else {
+		// if rbd create returned EEXIST (17) the image is already there and we just need to map
+		if exitError, ok := err.(*exec.ExitError); ok {
+			imageSpec := strings.Split(v.Name, "/") // strip the pool from the name
+			imageName := imageSpec[len(imageSpec)-1]
+			waitStatus := exitError.Sys().(syscall.WaitStatus)
+			if waitStatus.ExitStatus() == 17 || strings.Contains(stderr.String(), fmt.Sprintf("rbd image %s already exists", imageName)) {
+				log.Infof("Found existing Ceph volume='%s' - error=%s", v.Name, err)
+			} else {
+				msg := fmt.Sprintf("Failed to create Ceph volume='%s'. stderr=%s. exitcode=(%d) ", v.Name, stderr.String(), waitStatus.ExitStatus())
+				log.Errorf(msg)
+				return "", errors.New(msg)
+			}
+		} else {
+			log.Errorf(fmt.Sprintf("Failed to get exit code from ceph volume creation (volume='%s') - error=%s", v.Name, stderr.String()))
+			return "", err
+		}
+	}
+	if err := v.mapCephVolume(); err != nil {
+		return "", err
+	}
+	return v.Device, nil
+}
+
+func (v *Volume) LUKSEncryption() error {
+	luksDevMapperName := getLuksDeviceMapperName(v.Name)
+	cmd := exec.Command("cryptsetup", "luksOpen", "--allow-discards", "--key-file=-", v.Device, luksDevMapperName)
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		log.Errorf("Failed to luksOpen Ceph volume='%s' (device=%s) - error=%s", v.Name, v.Device, err)
+		return err
+	}
+	defer stdin.Close()
+	if err := cmd.Start(); err != nil {
+		log.Errorf("Failed to luksOpen Ceph volume='%s' (device=%s) - error=%s", v.Name, v.Device, err)
+		return err
+	}
+
+	key, err := getLuksKey(v.Name)
+	if err != nil {
+		log.Errorf("Failed to luksOpen Ceph volume='%s' (device=%s) - error=%s", v.Name, v.Device, err)
+		return err
+	}
+
+	io.WriteString(stdin, key)
+	stdin.Close()
+	if err := cmd.Wait(); err != nil {
+		log.Errorf("Failed to luksOpen Ceph volume='%s' (device=%s) - error=%s", v.Name, v.Device, err)
+		return err
+	}
+
+	v.LUKSDevice = path.Join(LuksDevMapperPath, luksDevMapperName)
+	return nil
+}
+
+func (v *Volume) Unmap(id string) error {
+	v.m.Lock()
+	defer v.m.Unlock()
+	defer v.unmapCephVolume()
+	fsType, err := fs.Detect(v.Device)
+	if err != nil {
+		return err
+	}
+	if fsType == cryptoLuksFsType {
+		luksDevMapperName := getLuksDeviceMapperName(v.Name)
+		cmd := exec.Command("cryptsetup", "luksClose", luksDevMapperName)
+		if err := cmd.Run(); err != nil {
+			log.Errorf("Failed to luksClose Ceph volume='%s' (device=%s) - error=%s", v.Name, v.Device, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func getLuksKey(name string) (string, error) {
+	return name, nil
```

Comment:
This needs to be encoded as the device name could be pool/name, could just pass luksDevMapperName to getLuksKey instead of v.Name.